src/uu/echo/src/echo.rs

```diff
@@ -21,73 +21,71 @@ mod options {
     pub const DISABLE_BACKSLASH_ESCAPE: &str = "disable_backslash_escape";
 }
 
-fn parse_code(
-    input: &mut Peekable<Chars>,
-    base: u32,
-    max_digits: u32,
-    bits_per_digit: u32,
-) -> Option<char> {
-    let mut ret = 0x8000_0000;
-    for _ in 0..max_digits {
-        match input.peek().and_then(|c| c.to_digit(base)) {
-            Some(n) => ret = (ret << bits_per_digit) | n,
+/// Parse the numeric part of the `\xHHH` and `\0NNN` escape sequences
+fn parse_code(input: &mut Peekable<Chars>, base: u8, max_digits: u32) -> Option<char> {
+    // All arithmetic on `ret` needs to be wrapping, because octal input can
+    // take 3 digits, which is 9 bits, and therefore more than what fits in a
+    // `u8`. GNU just seems to wrap these values.
+    // Note that if we instead make `ret` a `u32` and use `char::from_u32` will
+    // yield incorrect results because it will interpret values larger than
+    // `u8::MAX` as unicode.
+    let mut ret = input.peek().and_then(|c| c.to_digit(base as u32))? as u8;
+
+    // We can safely unwrap because we just peeked it.
+    input.next().unwrap();
+
+    for _ in 1..max_digits {
+        match input.peek().and_then(|c| c.to_digit(base as u32)) {
+            Some(n) => ret = ret.wrapping_mul(base).wrapping_add(n as u8),
             None => break,
         }
         input.next();
```

Comment:
even though Option isn't marked #[must_use], it might be nice to do `let _ = input.next()`

Reply:
though maybe clippy complains because it's *technically not necessary*, I always like it cause it's clear we don't care (instead of the unwrap above, which I guess could be the same `let _ =`

Reply:
at least make 'em the same

Reply:
Agreed and fixed!